doc/website/advanced/best-practice-for-testing.md

```diff
@@ -0,0 +1,263 @@
+    {
+        return m_number + rhs.m_number;
+    }
+
+    constexpr SingleDigitNumber operator-(const SingleDigitNumber rhs) const noexcept
+    {
+        return m_number - rhs.m_number;
+    }
+
+  private:
+    uint64_t m_number;
+};
+```
+
+This test fixture will be used
+
+```c++
+class SingleDigitNumber_test : public Test
+{
+  public:
+    void SetUp() override{};
+    void TearDown() override{};
+};
+```
+
+## First Attempt
+
+Well, this is quite a simple class, so the tests must also be simple, right?
+
+```c++
+TEST_F(SingleDigitNumber_test, TestClass)
+{
+    SingleDigitNumber number1;
+    SingleDigitNumber number2(3U);
+    SingleDigitNumber number3(2U);
+
+    auto number4 = number2 + number3;
+    auto number5 = number2 - number3;
+
+    EXPECT_TRUE(number4 == 5U);
+    EXPECT_EQ(1U, static_cast<uint64_t>(number5));
+}
+```
+
+Done and we reached 100% coverage. We can clap ourselves on the shoulders and move on. Well, except that we can't.
+
+The test above has several major and minor flaws. The first thing that leaps to the eye, it doesn't follow the **AAA** pattern. When the test fails, we don't know why and have to look in the code to figure out what went wrong. But even if only one aspect of the class would have been tested, there would still be essentially just a `[  FAILED  ] SingleDigitNumber_test.TestClass` output and we would need to look at the code to know what exactly is broken. This test case checks too many aspects of the code and doesn't have a sensible name.
+
+Furthermore, the default constructor is called in the test, but it is never checked to do the right thing. If there were a check, it would have revealed that `m_number` is not correctly initialized.
+
+Then, `EXPECT_TRUE` is used to compare the values. While this works, when the test fails we just get the information of the failure, but not which values were compared.
+To get this information `EXPECT_EQ` or `EXPECT_THAT` should be used. Furthermore, it's important to easily distinguish between the actual and the expected value, therefore the ordering of the values should be the same as if `EXPECT_THAT` would be used. The first value should be the actual value and the second one the expected.
+
+While the coverage might be 100%, there are no tests for:
+- invalid parameters, e.g. passing `10` to the constructor
+- overflow with `operator+`, e.g. adding `7` and `8`
+- underflow with `operator-`, e.g. subtracting `8` from `7`
+
+Here, ZOMBIES comes into play and gives us some guide to identify this cases.
+
+Some of this test might made it necessary to define the behavior of the class in the first place, e.g. defining that an `operator+` will saturate to the max value instead of overflowing.
+
+## How To Do It Better
+
+At first, the test is split into multiple test cases according the AAA pattern
+
+```c++
+TEST_F(SingleDigitNumber_test, DefaultConstructedObjectIsCorrectlyInitialized)
```

Comment:
If there is no complex initialization needed and the sut initialization is not universal to each tests we could also use TEST without fixtures. Probably not a good idea to advocate though, we can do useful things in the fixture like suppressing output by capturing ``std::cout`` and so on.

It should almost always be preferred to use the fixture, as you demonstrate (but tests without should be still allowed).